core/src/main/java/org/infinispan/marshall/core/internal/InternalMarshaller.java

```diff
@@ -0,0 +1,227 @@
+      // Ignore length since boundary checks are not so useful here where the
+      // unmarshalling code knows what to expect specifically. E.g. if reading
+      // a byte[] subset within it, it's always appended with length.
+      BytesObjectInput in = BytesObjectInput.from(buf, offset, this);
+      return objectFromObjectInput(in);
+   }
+
+   @Override
+   public Object objectFromInputStream(InputStream is) throws IOException, ClassNotFoundException {
+      // This is a very limited use case, e.g. reading from a JDBC ResultSet InputStream
+      // So, this copying of the stream into a byte[] has not been problematic so far,
+      // though it's not really ideal.
+      int len = is.available();
+      ExposedByteArrayOutputStream bytes;
+      byte[] buf;
+      if(len > 0) {
+         bytes = new ExposedByteArrayOutputStream(len);
+         buf = new byte[Math.min(len, 1024)];
+      } else {
+         // Some input stream providers do not implement available()
+         bytes = new ExposedByteArrayOutputStream();
+         buf = new byte[1024];
+      }
+      int bytesRead;
+      while ((bytesRead = is.read(buf, 0, buf.length)) != -1) bytes.write(buf, 0, bytesRead);
+      return objectFromByteBuffer(bytes.getRawBuffer(), 0, bytes.size());
+   }
+
+   @Override
+   public boolean isMarshallable(Object o) throws Exception {
+      Class<?> clazz = o.getClass();
+      boolean containsMarshallable = marshallableTypeHints.isKnownMarshallable(clazz);
+      if (containsMarshallable) {
+         boolean marshallable = marshallableTypeHints.isMarshallable(clazz);
+         if (trace)
+            log.tracef("Marshallable type '%s' known and is marshallable=%b",
+                  clazz.getName(), marshallable);
+
+         return marshallable;
+      } else {
+         if (isMarshallableCandidate(o)) {
+            boolean isMarshallable = true;
+            try {
+               objectToBuffer(o);
+            } catch (Exception e) {
+               isMarshallable = false;
+               throw e;
+            } finally {
+               marshallableTypeHints.markMarshallable(clazz, isMarshallable);
+            }
+            return isMarshallable;
+         }
+         return false;
+      }
+   }
+
+   private boolean isMarshallableCandidate(Object o) {
+      return o instanceof Serializable
+            || externalizers.marshallable(o).isMarshallable()
+            || o.getClass().getAnnotation(SerializeWith.class) != null
+            || isExternalMarshallable(o);
+   }
+
+   private boolean isExternalMarshallable(Object o) {
+      try {
+         return external.isMarshallable(o);
+      } catch (Exception e) {
+         throw new NotSerializableException(
+               "Object of type " + o.getClass() + " expected to be marshallable", e);
+      }
+   }
+
+   @Override
+   public BufferSizePredictor getBufferSizePredictor(Object o) {
```

Comment:
Ideally I'd like to also remove the buffer size predictor for internal objects, and instead ask the objects themselves (or their externalizers) how big a buffer they think they need. But that's probably out of scope for this PR :)